### ISSUE TITLE
HTCondor support

### ISSUE THREAD

**Issue by szs8** [dask/dask-jobqueue#100]
Does this project have plans for HTCondor support?

**Comment by mrocklin** [comment]
cc @jrbourbeau

**Comment by jhamman** [comment]
I don't think we have any current plans to implement an HTCondorCluster but it does seem in scope for this project. If an interested developer wanted to give it a try, I suspect it would be relatively straightforward. That said, I don't have any experience with HTCondor so take my measure of "straightforward" with a grain of salt.

**Comment by mrocklin** [comment]
I think that @jrbourbeau has a prototype that he's testing on an in-house
cluster

**Comment by jrbourbeau** [comment]
Sorry for the delayed response. @mrocklin is correct, I'm currently working on an `HTCondorCluster` implementation to add to dask-jobqueue and plan to open up a PR soon.

**Comment by szs8** [comment]
@jrbourbeau That's great! Looking forward to it.

**Comment by guillaumeeb** [comment]
@jrbourbeau any update? Do you have something to share? Maybe @szs8 would like to take a look at this?

**Comment by jrbourbeau** [comment]
Apologies for letting this linger. I got side-tracked with other work and haven't come back to this yet. @szs8 if you, or anyone else, are interested in contributing here, please feel free to take over.

**Comment by mivade** [comment]
@jrbourbeau Where does the existing implementation live?

**Comment by guillaumeeb** [comment]
@jrbourbeau do you have any prototype of HTCondorCluster to share, as mentioned above?

**Comment by guillaumeeb** [comment]
If anyone with access to an HTCondor cluster give it a try, I'd be happy to help!

**Comment by simone-codeluppi** [comment]
Hi
our lab has a small cluster with 12 nodes that is managed by HTCondor. 
I made a script for submitting the jobs and starting a dask cluster but it is not flexible and doesn't allow scaling  etc... just submit jobs for scheduler and workers and start a processing job. @guillaumeeb I will be happy to run some tests! Thank you!

**Comment by djhoese** [comment]
I was just asked by a supervisor about this compatibility. Google searches brought me to @matyasselmeci and their repository [here](https://github.com/matyasselmeci/dask_condor). I'm hoping they or someone else can give another status update.

**Comment by matyasselmeci** [comment]
Hi @djhoese and others,

My project was written to work with dask/distributed instead of dask/dask-jobqueue so some work would have to be done to adapt it.

My code differs from the other from the other batch system interfaces primarily because we wanted it to work without a shared filesystem, by having the user build a special tarball (using [build-worker-tarball](https://github.com/matyasselmeci/dask_condor/blob/master/build-worker-tarball) or [build-worker-tarball-conda](https://github.com/matyasselmeci/dask_condor/blob/master/build-worker-tarball-conda)) and having the worker jobs use HTCondor file transfer to transfer it.

The main difficulty I ran into in the end was that I couldn't figure out a way of building the worker tarball that was sufficiently user-friendly; because of this and because of a lack of interest in our department, the project fizzled.

I don't have time to work on the project anymore but I'm happy to answer any questions if someone wants to use the code as a starting point...

**Comment by djhoese** [comment]
@matyasselmeci That's too bad. Do you have any idea how much work would be needed to migrate your distributed PR to jobqueue? What if you assumed that users had a shared/networked file system?

**Comment by matyasselmeci** [comment]
I haven't looked at jobqueue at all, so I don't know how much the interface is different from distributed. I can get you an estimate later this week.

Since it sounds like there's interest in this again, I'll talk to my PI about getting some time to work on it.

**Comment by djhoese** [comment]
@matyasselmeci That would be great. My "secret" goal is to possibly get a JupyterHub instance running on the University of Wisconsin clusters, or at least at the SSEC where I work. I assume you also work on campus? If you want to meet in person to discuss some of this stuff and my ideas, let me know.

**Comment by guillaumeeb** [comment]
@matyasselmeci or @simone-codeluppi, again I would be happy to help extract the relevant bits of your scripts to a _JobQueueCluster_ implementation for HTCondor. I think this should be quite easy if you've already run a Dask cluster using `dask-scheduler` or `dask-worker` commands wrapped into job scripts.

**Comment by matyasselmeci** [comment]
@guillaumeeb yes, it looks like the changes wouldn't be too severe. One thing is that dask_condor uses [HTCondor's Python bindings](https://pypi.org/project/htcondor/) (which include some compiled code) to submit and control jobs instead of command-line tools (as the other JobQueueClusters do), so it may have to stay a separate project.

@djhoese yes, I work at Comp Sci. My colleagues and I would be very interested in meeting with you and discussing your projects; mind if I contact you via email? (I found your SSEC email address in the UW directory, if that's a good one to use.)

**Comment by djhoese** [comment]
@matyasselmeci Yes that works.

**Comment by matyasselmeci** [comment]
I've started work on this and it shouldn't be too difficult; I can have a branch in dask_condor that's usable for external testing in a week or two. @mrocklin, if you're OK with adding an optional dependency on htcondor (`extras_require=dict(htcondor="htcondor>=8.6.0")` in setup.py) then I can turn it into a pull request after that.

**Comment by matyasselmeci** [comment]
I'm using JobQueueCluster as the base class; I'll try to make it as compatible as possible and will follow the conventions the other batch system implementations are using.

**Comment by jhamman** [comment]
@matyasselmeci - thanks for sticking around here and your continued interest. I have not personally used HTCondor so forgive me if my questions are off base a bit. I'm mostly interested in discussing if there is value in have the `HTCondorCluster` use the the same machinery that the other cluster managers do. I quickly looked at [these](https://research.cs.wisc.edu/htcondor/manual/quickstart.html) docs and it seems this should be quite easy to do.

It seems we could basically template out a job script (just like in the PBSCluster) that looks like: 

```
# /tmp/tempfile.sub -- run one instance of dask_worker
executable              = %(python)s -m distributed.cli.dask_worker
arguments               = "%(worker_command)s"
log                     = %(log_directory)/dask-worker-log.txt
queue
```

and then the submit/cancel commands:

```python
    submit_command = 'condor_submit'
    cancel_command = 'condor_rm'
```

It seems like additional arguments are possible in the job script to specify wall time, resources, etc. 

Now, I could be missing a fundemental part of HTCondor so please correct me if I'm wrong here.

**Comment by matyasselmeci** [comment]
Little more complicated than that; `executable` needs to be a single file (whatever happened to the `dask-worker` executable btw?); the quoting for arguments is... nonstandard...; `log` is actually the record of what the HTCondor scheduler does to the job, not the dask-worker's output.

HTCondor is used for submitting multiple jobs at the same time so the job ID has the form `13.5` where the `13` is the ID of the group of jobs submitted ("ClusterId" -- bad name but it's historical) and the `5` is the specific job within that group ("ProcId" -- same here); `condor_submit` only gives you the ClusterId in the output (but the ProcIds are 0..n-1 as expected)...

Basically the devil is in the details. It's possible to do it that way but you'll need to do a lot of pre-processing of the parameters anyway or else it won't behave the way you expect it to.

**Comment by guillaumeeb** [comment]
I've no strong opinion here. I would prefer to have the same scheme for all implementations as said before, but I'm happy if we can have a different working solution, that still fits in dask-jobqueue.

If you've started work on this and have some interest in it, I say keep going and ping us as soon as you have something thats understandable so that we can give feedback!

**Comment by lesteve** [comment]
Just a note: it may be useful to look at `IPython.parallel` implementation for `HTCondor`:
https://github.com/ipython/ipyparallel/blob/beb400fd87f59504f231baa28c9b89ea76ab4f79/ipyparallel/apps/launcher.py#L1436

**Comment by guillaumeeb** [comment]
Closed by #245.
